### PR TITLE
Support archive traces for ES storage

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves https://github.com/jaegertracing/jaeger/issues/818
Related to https://github.com/jaegertracing/jaeger/issues/1242

This adds the implementation of archive storage for ES. There are two possible configurations how to use this feature:

* the default - it will store and read archived traces from `jaeger-archive-span` index.
* using rollover API - it writers spans to an alias `jaeger-span-archive-write` and reads from an alias `jaeger-span-archive-read`. This deployment requires an external component that performs index managent: creating index, creating alias, callig rollover API and managing `max-span-age` by removing indices from read alias. This component is `esRollover.py` available as `jaeger-es-rollover` docker image.

As the default index name does not collide with aliases used by rollover users can seameasly migrate from default deployment to rollover by putting  `jaeger-archive-span` index into the read alias.

### Default use case
It does not require any configuration just deploying jaeger.

### Rollover use case
It requires to create index `jaeger-span-archive-000001` and aliases `jaeger-span-archive-read`, `jaeger-span-archive-write`.

To enable rollover the following switch has to be enabled
```
--es-archive.use-aliases                                    Use read and write aliases for indices. Use this option with Elasticsearch rollover API. It requires an external component to create aliases before startup and then performing its management. Note that es-archive.max-span-age is not taken into the account and has to be substituted by external component managing read alias. (default false)
```

#### Initialization
Initialize - create index and aliases
```
ARCHIVE=true UNIT=seconds python3 esRollover.py init localhost:9200
```

#### Rollover
Now we can start jaeger and archive traces. The next step is to rollover to new write index:
```
ARCHIVE=true CONDITIONS='{"max_age": "1s"}'  python3 esRollover.py  rollover localhost:9200
```
I am using max age 1s to trigger the rollover immmediately. This step can be repeated multiple times, The read alias should always point to all created indices, whereas write alias only to the last one. 

#### Managing max-span-age and delete old indices
The last step is to remove old indices from read alias - to mimic `max-span-age`.
```
ARCHIVE=true  UNIT=minutes UNIT_COUNT=2  python3 esRollover.py  lookback localhost:9200
```
We can also delete the old indices entirely by running es-index-cleaner
```
ARCHIVE=true python3 esCleaner.py 1 localhost:9200
```
Note that it supports only days at unit.

#### Other
Indices: http://localhost:9200/_cat/indices
Aliases: http://localhost:9200/_cat/aliases

UI config to enable archive tab: `--query.ui-config ui-config.json`
```
{
    "archiveEnabled": true
}
```

ES 6.4.4 supports `is_write_index` - it means that a single alias can be used for reads and writes (it points to multiple indices from which only one is write). We can add support later by using a different flag.

The following command can be used to rollover and put the new index to read alias.
```bash
curl -ivX PUT -H "Content-Type: application/json" localhost:9200/jaeger-span-archive-000001 -d '{
  "aliases": {
    "jaeger-span-archive-write": {"is_write_index": true},
    "jaeger-span-archive-read": {}
  }
}'
```

Note that all actions provided by `esRollover.py` can be written declarativelly as yml files for curator. Here is a working examlple https://gist.github.com/pavolloffay/5ce3f59f0967aadd5b019be411121c6c. The problem is to support index prefix. Yaml approach is quite nice but does not satifly our requirements if we want to have some fields configurable.

### CODE REVIEW

File: plugin/storage/es/factory.go [hunk 1]
```diff
@@ -133,3 +141,42 @@ func loadTagsFromFile(filePath string) ([]string, error) {
 	}
 	return tags, nil
 }
+
+// CreateArchiveSpanReader implements storage.ArchiveFactory
+func (f *Factory) CreateArchiveSpanReader() (spanstore.Reader, error) {
```
Comment (yurishkuro) [inline plugin/storage/es/factory.go]: can we DRY this to be written only once based on the passed cfg?

File: plugin/storage/es/spanstore/writer.go [hunk 2]
```diff
@@ -104,20 +106,28 @@ func NewSpanWriter(p SpanWriterParams) *SpanWriter {
...
 
 // WriteSpan writes a span and its corresponding service:operation in ElasticSearch
 func (s *SpanWriter) WriteSpan(span *model.Span) error {
-	spanIndexName := indexWithDate(s.spanIndexPrefix, span.StartTime)
-	serviceIndexName := indexWithDate(s.serviceIndexPrefix, span.StartTime)
+	var spanIndexName string
+	if s.archive {
+		spanIndexName = archiveIndex(s.spanIndexPrefix)
+	} else {
+		spanIndexName = indexWithDate(s.spanIndexPrefix, span.StartTime)
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/writer.go]: seems like if we merged with PR with index name pattern, we would not have to make these splits between archive/non-archive.

Comment (pavolloffay) [reply plugin/storage/es/spanstore/writer.go]: yes this is a bit ugly. We can set the func in the constructor.

File: plugin/storage/es/factory.go (hunk 1 above)
Comment (pavolloffay) [reply plugin/storage/es/factory.go]: yes this is just a WIP

**Comment by pavolloffay** [comment]
This is ready for review @jaegertracing/elasticsearch @yurishkuro @objectiser 

Please read the first comment. We could also automatically default to write and read aliases and do not support no ttl use case.

The question is also who should create aliases. We could embed it in Jaeger, although I am not sure about that approach.

File: plugin/storage/es/spanstore/index_utils.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanstore
+
+import (
+	"time"
+)
+
+type spanAndServiceIndexFce func(spanTime time.Time) (string, string)
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/index_utils.go]: what does Fce stand for?

**Comment by pavolloffay** [comment]
> Do you think the same is possible to achieve with ES, i.e. by parameterizing the reader/writer? 

That is what I wanted to achieve here too. The archive parameter is not present in reader/writer methods. it's only passed to the constructor which changes how indices are resolved.

File: plugin/storage/es/spanstore/index_utils.go (hunk 3 above)
Comment (pavolloffay) [reply plugin/storage/es/spanstore/index_utils.go]: abbreviation for function in Czech language

File: plugin/storage/es/spanstore/reader.go [hunk 4]
```diff
@@ -96,6 +96,7 @@ type SpanReader struct {
 	spanIndexPrefix         string
 	serviceIndexPrefix      string
 	spanConverter           dbmodel.ToDomain
+	indicesForTimeRangeFce  indicesForTimeRangeFn
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/reader.go]: s/indicesForTimeRangeFce/indicesForTimeRangeFn/

**Comment by yurishkuro** [comment]
>>Do you think the same is possible to achieve with ES, i.e. by parameterizing the reader/writer?

>That is what I wanted to achieve here too. The archive parameter is not present in reader/writer methods. it's only passed to the constructor which changes how indices are resolved.

That's exactly what concerns me: imho nothing in `plugin/storage/es/*` should have any references to archive/non-archive names, we're only half-way there. The storage can support two modes: daily indices and index aliases. The distinction between archive/non-archive ideally should be consolidated in one place, and the outcome of that one place is a config object that says these are the names/patterns and the daily vs. alias mode. Currently too many places in the code know about archive/non-archive.

Another high-level question: is it possible to transparently support the daily-index mode via index-alias mode (I assume not)?

**Comment by masteinhauser** [comment]
> is it possible to transparently support the daily-index mode via index-alias mode (I assume not)?

This is possible, at least for reads, via the curator and assigning multiple indices to the same alias.


> For the example above, a glob pattern can also be used to associate an alias to more than one index that share a common name:
```
POST /_aliases
{
    "actions" : [
        { "add" : { "index" : "test*", "alias" : "all_test_indices" } }
    ]
}
```
> In this case, the alias is a point-in-time alias that will group all current indices that match, it will not automatically update as new indices that match this pattern are added/removed.

Again, Aliases work well (and transparently) only for read operations:
> It is an error to index to an alias which points to more than one index.
However, I believe @pavolloffay mentioned (perhaps elsewhere) of [specifying the single write-index to use while still behind a single alias](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html#aliases-write-index).

Ref: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html

**Comment by yurishkuro** [comment]
It would be nice if jaeger collector/ingester would perform the management of the write indices. It already creates new daily index opportunistically. We have a simple master election implementation for Cassandra used by adaptive sampling, can probably extend it to ES to pick one of the nodes to perform index rollover.

**Comment by pavolloffay** [comment]
@yurishkuro what do you mean by:

> It would be nice if jaeger collector/ingester would perform the management of the write indices

Do you mean to create aliases if the rollover is being used? I think that is the only index functionality we could manage from jaeger. The other API that is necessary to call is `localhost:9200/jaeger-span-archive/_rollover`, however it does not make sense to call it from collector because:
* the body accepts various conditions - it does not make sense to pollute our API to accept it
* the call can fail and the component calling it should act accordingly - this is more of operator's job.

We will have to provide a component to create aliases and call rollover API. The first comment contains the curl calls. If we embed the alias creation to the collector it limits the naming patterns (maybe something more). I would put the index management into a separate component it's more flexible, power users could skip it and use their configurations.

Note that the default behavior is without rollover API. What we have to solve is the migration from single index to rollover. If we will use slightly different index names for rollover we are fine users can easily migrate by putting old index to read alias. The other option is to use reindex API to change archive index name and then create alias with the original name. However reindex copies data which takes time so there will be downtime.

**Comment by yurishkuro** [comment]
> Do you mean to create aliases if the rollover is being used?

Yes, it's what (I think) I meant. Sorry, too many nuances going on here.

+1 on having an external mgmt component outside of collectors. However, are you sure it would work with daily indices, even if we normalize them with aliases? Today we create a new index when we see a span with the timestamp for the next day, right? Mgmt component won't have that info.

Also, is es curator difficult to set up? If not, what is the value of writing our own? If we did mgmt out of collectors it's attractive because there are fewer things to deploy (btw totally agree with/ your point about the API for such thing).

**Comment by pavolloffay** [comment]
>  However, are you sure it would work with daily indices, even if we normalize them with aliases? 

The alias does not contain date it would be simply an alias `jaeger-span`. This alias would point to an active write index `jaeger-span-00001`. The external component would call rollover to create a new write index if some conditions are valid eg. size, date or number of shards. This also solves the current problem when writer creates index too far in the future https://github.com/jaegertracing/jaeger/issues/841.

**Comment by pavolloffay** [comment]
@jaegertracing/elasticsearch @yurishkuro @jpkrohling @objectiser could you please review?

While doing the review could you please deploy multiple jaeger instances to single ES and especially verify that index prefix works as expected.

**Comment by objectiser** [comment]
@pavolloffay Only done some initial steps currently - set up the normal (non-rollover) indexes did some requests, then initialised using the rollover script using the same `INDEX_PREFIX` value as used for the non-rollover archive prefix. Then ran a request but now getting the following after I try viewing the traces:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xaddcb3]

goroutine 16 [running]:
github.com/jaegertracing/jaeger/pkg/es/config.(*Configuration).NewClient.func2(0x1, 0xc000a67220, 0x1, 0x1, 0x0, 0x15acaa0, 0xc0002e1840)
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/pkg/es/config/config.go:109 +0xe3
github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic%2ev5.(*bulkWorker).commit(0xc0000483c0, 0x15b6180, 0xc0000b8030, 0x1, 0xc000400001)
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic.v5/bulk_processor.go:506 +0x415
github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic%2ev5.(*bulkWorker).work(0xc0000483c0, 0x15b6180, 0xc0000b8030)
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic.v5/bulk_processor.go:455 +0x283
created by github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic%2ev5.(*BulkProcessor).Start
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic.v5/bulk_processor.go:299 +0x184
```

File: plugin/storage/es/factory.go [hunk 5]
```diff
@@ -33,6 +33,11 @@ import (
 	"github.com/jaegertracing/jaeger/storage/spanstore"
 )
 
+const (
+	primaryNamespace = "es"
+	archiveNamespace = "es-archive"
```
Comment (objectiser) [inline plugin/storage/es/factory.go]: Can we change this to `es.archive` - reason being that the operator currently filters out storage specific options based on the first part ('.' separated) - and "es" is used for elasticsearch options.

Comment (pavolloffay) [reply plugin/storage/es/factory.go]: Isn't it an implementation detail of the operator?

I would like to be rather consistent with other storages configurations in Jaeger. Cassandra flags:
```
--cassandra-archive.tls
--cassandra.tls
```

**Comment by pavolloffay** [comment]
> Then ran a request but now getting the following after I try viewing the traces:

@objectiser I have followed your instructions but the exception didn't happen. Could you please provide an exat reproducer? Also the stack trace you pasted here is from collector - es bulk processor and not from query service.

**Comment by objectiser** [comment]
@pavolloffay When using the loopback command, if there are no indices to remove it currently throws:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/curator/actions.py", line 86, in remove
    ilo.empty_list_check()
  File "/usr/local/lib/python3.7/site-packages/curator/indexlist.py", line 228, in empty_list_check
    raise exceptions.NoIndices('index_list object is empty.')
curator.exceptions.NoIndices: index_list object is empty.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/es-rollover/esRollover.py", line 109, in <module>
    main()
  File "/es-rollover/esRollover.py", line 51, in main
    read_alias_lookback(client, write_alias, read_alias, os.getenv('UNIT', UNIT), int(os.getenv('UNIT_COUNT', UNIT_COUNT)))
  File "/es-rollover/esRollover.py", line 99, in read_alias_lookback
    alias.remove(ilo)
  File "/usr/local/lib/python3.7/site-packages/curator/actions.py", line 98, in remove
    raise exceptions.NoIndices('No indices to remove from alias')
curator.exceptions.NoIndices: No indices to remove from alias
```
I think it would be better if it just logs an info or warning message?

File: plugin/storage/es/factory.go (hunk 5 above)
Comment (objectiser) [reply plugin/storage/es/factory.go]: Yeah that is fine - was just thinking if only associated with elasticsearch options, then using a single part would help - but can be accommodated in the operator.

**Comment by objectiser** [comment]
@pavolloffay will try to reproduce from a clean start, but it was definitely from the query service:
```
$ SPAN_STORAGE_TYPE=elasticsearch ./query-linux --es.index-prefix=prod --es-archive.index-prefix=prod-archive --es-archive.use-aliases --query.ui-config=ui-config.json
{"level":"info","ts":1547654902.4261525,"caller":"healthcheck/handler.go:99","msg":"Health Check server started","http-port":16687,"status":"unavailable"}
{"level":"info","ts":1547654902.6611705,"caller":"query/main.go:134","msg":"Registering metrics handler with HTTP server","route":"/metrics"}
{"level":"info","ts":1547654902.661216,"caller":"healthcheck/handler.go:133","msg":"Health Check state change","status":"ready"}
{"level":"info","ts":1547654902.6612267,"caller":"query/main.go:143","msg":"Starting HTTP server","port":16686}
{"level":"error","ts":1547654921.427332,"caller":"zap/logger.go:33","msg":"error when flushing the buffer: write udp 127.0.0.1:55827->127.0.0.1:6831: write: connection refused","stacktrace":"github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/log/zap.(*Logger).Error\n\t/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/log/zap/logger.go:33\ngithub.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go.(*remoteReporter).processQueue.func1\n\t/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/reporter.go:257\ngithub.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go.(*remoteReporter).processQueue\n\t/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/reporter.go:267"}
{"level":"error","ts":1547654960.4273534,"caller":"zap/logger.go:33","msg":"error when flushing the buffer: write udp 127.0.0.1:55827->127.0.0.1:6831: write: connection refused","stacktrace":"github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/log/zap.(*Logger).Error\n\t/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/log/zap/logger.go:33\ngithub.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go.(*remoteReporter).processQueue.func1\n\t/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/reporter.go:257\ngithub.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go.(*remoteReporter).processQueue\n\t/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/github.com/uber/jaeger-client-go/reporter.go:267"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xaddcb3]

goroutine 16 [running]:
github.com/jaegertracing/jaeger/pkg/es/config.(*Configuration).NewClient.func2(0x1, 0xc000a67220, 0x1, 0x1, 0x0, 0x15acaa0, 0xc0002e1840)
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/pkg/es/config/config.go:109 +0xe3
github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic%2ev5.(*bulkWorker).commit(0xc0000483c0, 0x15b6180, 0xc0000b8030, 0x1, 0xc000400001)
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic.v5/bulk_processor.go:506 +0x415
github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic%2ev5.(*bulkWorker).work(0xc0000483c0, 0x15b6180, 0xc0000b8030)
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic.v5/bulk_processor.go:455 +0x283
created by github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic%2ev5.(*BulkProcessor).Start
	/home/gbrown/repositories/go/src/github.com/jaegertracing/jaeger/vendor/gopkg.in/olivere/elastic.v5/bulk_processor.go:299 +0x184
```

**Comment by objectiser** [comment]
@pavolloffay Steps to reproduce:
```
docker run -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" elasticsearch:5.6
```
* Checked indices and aliases - both empty
```
SPAN_STORAGE_TYPE=elasticsearch ./query-linux --es.index-prefix=prod --es-archive.index-prefix=prod-archive --query.ui-config ui-config.json
```
```
SPAN_STORAGE_TYPE=elasticsearch ./collector-linux --es.index-prefix=prod --es-archive.index-prefix=prod-archive
```
* Generate first traced request
* View trace instance in UI and archive
```
docker run -it --rm --net=host -e ARCHIVE=true -e UNIT=seconds -e INDEX_PREFIX=prod-archive jaegertracing/jaeger-es-rollover init localhost:9200
```
At this point, aliases show:
```
$ curl http://localhost:9200/_cat/aliases?v
alias                                  index                                   filter routing.index routing.search
prod-archive:jaeger-span-archive-read  prod-archive:jaeger-span-archive-000001 -      -             -
prod-archive:jaeger-span-archive-write prod-archive:jaeger-span-archive-000001 -      -             -
prod-archive:jaeger-span-archive-read  prod-archive:jaeger-span-archive        -      -             -
prod-archive:jaeger-span-archive-write prod-archive:jaeger-span-archive        -      -             -
```
* Restart query and collector with:
```
SPAN_STORAGE_TYPE=elasticsearch ./query-linux --es.index-prefix=prod --es-archive.index-prefix=prod-archive --es-archive.use-aliases --query.ui-config=ui-config.json
```
```
SPAN_STORAGE_TYPE=elasticsearch ./collector-linux --es.index-prefix=prod --es-archive.index-prefix=prod-archive --es-archive.use-aliases
```
* Generated second traced request
* View trace instance in UI and archive
* At this point I went back to the search page in the UI and find traces, but then about 10 seconds later saw the panic and query service quit. It is not instantaneous.

**Comment by pavolloffay** [comment]
@objectiser thanks I was able to reproduce. The problem is that write alias points to two indices so when you archive a trace in UI query will use bullk processor to store it - and it fails. I will fix the script.

File: plugin/storage/es/factory.go (hunk 5 above)
Comment (pavolloffay) [reply plugin/storage/es/factory.go]: Ok then I am resolving this, we will handle it in the operator.

**Comment by pavolloffay** [comment]
> When using the loopback command, if there are no indices to remove it currently throws:

I have made it more robust and added output when it does something

**Comment by objectiser** [comment]
@pavolloffay The panic on query service is fixed. But still seeing a stack when doing the `lookback` (although seems smaller). Is this expected? Thought a warning message may be better?
```
$ docker run -it --rm --net=host -e ARCHIVE=true -e UNIT=minutes -e UNIT_COUNT=120 -e INDEX_PREFIX=ro-archive jaegertracing/jaeger-es-rollover lookback localhost:9200
GET http://localhost:9200/prod-archive%3Ajaeger-span-archive,prod-archive%3Ajaeger-span-archive-000001,prod%3Ajaeger-service-2019-01-17,prod%3Ajaeger-span-2019-01-17/_alias/ro-archive%3Ajaeger-span-archive-read [status:404 request:0.005s]
Traceback (most recent call last):
  File "/es-rollover/esRollover.py", line 128, in <module>
    main()
  File "/es-rollover/esRollover.py", line 53, in main
    read_alias_lookback(client, write_alias, read_alias, os.getenv('UNIT', UNIT), int(os.getenv('UNIT_COUNT', UNIT_COUNT)))
  File "/es-rollover/esRollover.py", line 104, in read_alias_lookback
    ilo.filter_by_alias(aliases=[write_alias], exclude=True)
  File "/usr/local/lib/python3.7/site-packages/curator/indexlist.py", line 846, in filter_by_alias
    self.empty_list_check()
  File "/usr/local/lib/python3.7/site-packages/curator/indexlist.py", line 228, in empty_list_check
    raise exceptions.NoIndices('index_list object is empty.')
curator.exceptions.NoIndices: index_list object is empty.
```

**Comment by objectiser** [comment]
@pavolloffay My mistake - used wrong index - it works now.

**Comment by pavolloffay** [comment]
@jaegertracing/elasticsearch it would be great if any of you guys could have a look and briefly comment on our approach.

**Comment by pavolloffay** [comment]
It's important to mention that the similar approach would be used for the main indices when using rollover.

**Comment by pavolloffay** [comment]
I think we should go forward and merge this PR. It will be available in latest images so people can try. If nobody objects I will merge it today EOD.

File: plugin/storage/es/factory.go [hunk 6]
```diff
@@ -71,45 +79,22 @@ func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger)
...
-		Logger:            f.logger,
-		MetricsFactory:    f.metricsFactory,
-		NumShards:         f.primaryConfig.GetNumShards(),
-		NumReplicas:       f.primaryConfig.GetNumReplicas(),
-		IndexPrefix:       f.primaryConfig.GetIndexPrefix(),
-		AllTagsAsFields:   f.primaryConfig.GetAllTagsAsFields(),
-		TagKeysAsFields:   tags,
-		TagDotReplacement: f.primaryConfig.GetTagDotReplacement(),
-	}), nil
+	return createSpanWriter(f.metricsFactory, f.logger, f.archiveClient, f.primaryConfig, false)
```
Comment (ledor473) [inline plugin/storage/es/factory.go]: @pavolloffay I think the client here should be `f.primaryClient` and not `f.archiveClient`

**Pull request merged**